### PR TITLE
Review compilation steps documentation

### PR DESCRIPTION
**Pull request by fm-117**
This is only to review the COMPILATION.MD file, not to merge.

The final document will reside in our wiki but not in source control.

### CODE REVIEW

**Review by efr15 — CHANGES_REQUESTED** [review]

File: Documentation/COMPILATION_STEPS.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+- _Text Change Notification Handling_
+- _Text Update_
+- _Tokens Update_
+- _Tokens Preprocessing_
+- _Code Element Update_
+- _AST building_
+- _Cross check_
+- _Code analysis_
+
+The first five steps: _Text Change Notification Handling_, _Text Update_, _Tokens Update_, _Tokens Preprocessing_ and _Code Element Update_ are **incremental**. They are line-based, meaning that they operate at line level and use lines to define sections of code that are modified by a given text change.
```
[the remark this answers is not in the document]
Comment (fm-117) [reply Documentation/COMPILATION_STEPS.md:14]: done

File: Documentation/COMPILATION_STEPS.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+|Class|Method|Input|Output|
+|-|-|-|-|
+|`CompilationDocument`|`UpdateTokensLines`|List of `DocumentChange<ICobolTextLine>`|Updated `compilationDocumentLines`, list of `DocumentChange<ITokensLine>`|
+|`CompilationDocument`|`RefreshTokensDocumentSnapshot`|`void`|Updated `TokensDocumentSnapshot`|
+
+During this step, the lines that have been altered are scanned to identify their tokens. The `RefreshTokensDocumentSnapshot`  only updates the `TokensDocument` instance, the `UpdateTokensLines` is doing all the work by calling the `ScannerStep` class.
+
+To enable incremental scanning, the ScannerStep uses the `MultilineScanState` object. A line must be rescanned when either:
+- it is part of the modified lines (and consequently has no tokens yet)
+- it is part of a modified continuation group
```
[the remark this answers is not in the document]
Comment (fm-117) [reply Documentation/COMPILATION_STEPS.md]: done

File: Documentation/COMPILATION_STEPS.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+### ANTLR parsing
+
+The parsing itself is triggered through a single call to `CodeElementsParser.cobolCodeElements()`. This is within this operation that the tokens are actually iterated and the `COPY` / `REPLACE` directives are processed. To give ANTLR the real tokens, the iterator must:
+- import tokens from documents included by `COPY`s instructions
+  - applying the `REPLACING` clause if any
+- alter tokens that match current `REPLACE` directive
+
+### Building CodeElements
+
+ANTLR returns a single `CodeElementsParser.CobolCodeElementsContext`. Context objects are not kept in the final result but are rather converted into real `CodeElement`s objects. This is done through a visitor pattern, the visitor object is `CodeElementBuilder` which turns each context object into its corresponding CodeElement object and also gather tokens of this CE into its `ConsumedTokens` collection.
```
[the remark this answers is not in the document]
Comment (fm-117) [reply Documentation/COMPILATION_STEPS.md:160]: done

File: Documentation/COMPILATION_STEPS.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+By convention, structures allowing children are entered when the corresponding `StartXXX` method is called and exited when the `ExitXXX` method is called. Childless structures are visited with a `OnXXX` method call.
+
+For example:
+- `ProgramClassBuilder.StartDataDivision` signals the beginning of the `DATA DIVISION` of the program, after this call the current node is the newly created `DataDivision` node allowing to add data definitions children.
+- `ProgramClassBuilder.EndDataDivision` signals the end of the DATA DIVISION, thus the current node will be the program node.
+- `ProgramClassBuilder.OnGobackStatement` signals the encounter of a `GobackStatement`. The method creates the `Goback` node, enters it as a child of the current node (most probably the `ProcedureDivision` node) and exit immediately.
+
+The parsing errors produced by CUP parser are stored in the `Diagnostics` collection of the `TemporarySemanticDocument`.
+
```
[the remark this answers is not in the document]
Comment (fm-117) [reply Documentation/COMPILATION_STEPS.md:178]: done

**Review by efr15 — DISMISSED** [review]

File: Documentation/COMPILATION_STEPS.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+|-|-|-|-|
+|`CompilationUnit`|`RefreshCodeAnalysisDocumentSnapshot`|Up-to-date `ProgramClassDocumentSnapshot`|Updated `CodeAnalysisDocumentSnapshot`|
+
+This final step runs external analyzers on the latest `ProgramClassDocumentSnapshot`. To allow all types of code analysis, the analyzers are given all previous document snapshots (Tokens, ProcessedTokens, CodeElements, TemporaryProgramClass and ProgramClass).
+
+Each analyzer receives all five document snapshot in sequence through calls to their `Inspect` methods. The results of the analysis (mainly diagnostics) are stored in a new instance of `InspectedProgramClassDocument`.
+
+## About snapshots and result consistency.
+
+The parser does not use multi-threading to parallelize source code processing. However the client of the parser may choose to perform multiple operations on the same source simultaneously. Typically in LanguageServer, a thread may be reading results of an incremental compilation round while the timer thread is already performing the remaining non-incremental compilation steps.
```
Comment (smedilol) [inline Documentation/COMPILATION_STEPS.md:207]: Our LSP implementation is rather mono thread.
We have 3 threads:

- 1 thread to receive messages and enqueue them
- 1 thread to process messages  from the queue
- 1 thread / timer that just push an event in the queue.

So I'd rather say that everything was designed to allow multithreading, but currently we don't use it. And we have no proof that everything will works fine with multithreading.

**Review by smedilol — CHANGES_REQUESTED** [review]

File: Documentation/COMPILATION_STEPS.md (hunk 5 above)
Comment (fm-117) [reply Documentation/COMPILATION_STEPS.md]: Rewritten

**Review by smedilol — APPROVED** [review]

**Review by efr15 — APPROVED** [review]

**Comment by fm-117** [comment]
https://github.com/TypeCobolTeam/TypeCobol/wiki/Compilation-process updated.

Now closing this PR and deleting branch.

**Pull request closed without merging**